docs/design-proposals/tidb-stable-scheduling.md

```diff
@@ -0,0 +1,141 @@
+    + [Workarounds](#workarounds-1)
+
+## Glossy
+
+- Pod: A deployable object in Kubernetes.
+- Node: A node is worker machine to run pods in Kubernetes.
+- TiDB cluster: TiDB cluster is database cluster which is constructed by TiDB
+  server, PD server and TiKV server. Each TiDB/PD/TiKV server is a cluster too
+  which consist of one or more members.
+- TiDB server: One of key components of TiDB cluster. It's access point of TiDB
+  cluster.
+- TiDB member: member of TiDB server which holds a unique network identifier.
+- TiDB pod: for each TiDB member, there is at most one running pod. It may
+  crash or be replaced by a new pod.
+- Load balancer: A component which proxies traffic for applications, e.g. LVS,
+  HAProxy, F5, etc.
+
+## Motivation
+
+There are use cases that users need to access TiDB server from outside the
+Kubernetes cluster. But in some environments (e.g. on bare-metal machines), we
+may lack load balancer solution or need to configure IP addresses of TiDB
+services in existing load balancer.
+
+In this scenario, we need to use `NodePort` service. By default, NodePort
+services are available cluster-wide on all nodes but cannot propagate client's
+IP address to the end pods and may cause a second hop. 
+
+For propagating client's IP address and better performance, we can
+set `externalTrafficPolicy` of service to `Local`. A side-effect is the service of
+TiDB will be accessible only on the nodes which a running TiDB pod. To avoid
+manual intervention to update IP addresses in load balancer when performing
+a rolling update, we prefer to schedule new pod of TiDB member to its previous node.
+
+## Goals
+
+- Able to schedule new pod of TiDB member to its previous node
+
+### Non-Goals
+
+- Stable scheduling for TiKV/PD pods
+- Guarantee that new pod of TiDB member will be scheduled to its node if
+  there is another scheduler in cluster which may schedule pods to its node
+- Guarantee that new pod of TiDB member will be scheduler to its node if some
+  scheduling requirements changed
+
+## Use Cases
+
+### No need to update IP addresses of TiDB in load balancer outside of the Kubernetes cluster
+
+When a TiDB cluster is running in a dedicated Kubernetes cluster or nodes of
+TiDB cluster are reserved for it. After a rolling update of TiDB cluster is
+done, new pods of TiDB 
+
+## Proposal
+
+Currently, we have tidb-scheduler to schedule all pods of TiDB cluster. We can
+write a new predicate function for pods of TiDB server. In this new predicate
+function we can choose the its previous node of TiDB member it it exists in candidate
+nodes.
+
+## Implementation
+
+At first, we track assigned node of TiDB member in status of TiDBCluster.
+
+```
+type TiDBMember struct {
+  ...
+	// Node hosting pod of this TiDB member.
+	NodeName string `json:"node,omitempty"`
+}
+```
+
+In new predicate `StableScheduling`, we filter out other nodes for TiDB pod if
```

Comment:
Can I specify scheduling policy (whether enable `StableScheduling`) for a `TidbCluster` instance?

Reply:
Yes, but I'd like to add a global feature switch to control this scheduling behavior for all TiDB cluster instances tidb-operator managed. Because this is a best-effort policy, no harm if it fails.

What I am thinking is should we fail the pod (with a switch to control) if it cannot be scheduled to its previous node?

Reply:
One drawback is its previous node may not be the best node for the new node, e.g. there is another node which has more available CPU/Memory resources. However, whether to use this scheduling policy depends on cluster setup (use NodePort with Local policy or not), I think to control this behavior globally is enough. Furthermore, we can bypass this policy if TiDB service does not use `local` externalTrafficPolicy intelligently.

Reply:
I agree to add a switch for this scheduling policy as this feature is not suitable for all users.

Reply:
One way is to control globally: 

A flexible feature gate like this:

```
tidb-scheduler --features StableScheduling,FeatureA,FeatureB
```

It works like Kubernetes feature gate, but simpler (all disabled by default).

Or a dedicated flag

```
tidb-scheduler --enable-stable-scheduling
```

Another way is to control per TiDBCluster, add a field in TiDBCluster.

What do you prefer?

Reply:
I think it should be a global feature controlled by operator not `TidbCluster` CRD.

Reply:
I suppose both options are valuable:

* feature gate of scheduler: controlled by cluster admin, determine whether current cluster **is willing to support stable scheduling**.
* field of `TidbCluster` CRD: controlled by user who make the decision of whether using local mode and stable scheduling